### PR TITLE
fix for issue #50 (Add icmp OnTimeout callback support)

### PR DESCRIPTION
**Pull request by eunchong**
fix for issue #50 

Can use like this..

```Go
	pinger.OnTimeout = func(pkt *ping.Packet) {
		fmt.Printf("Request timeout for %s: icmp_seq=%d\n",
			pkt.IPAddr, pkt.Seq)
	}
```

### CODE REVIEW

File: ping.go [hunk 1]
```diff
@@ -497,6 +513,34 @@ type IcmpData struct {
...
+func (p *Pinger) checkWaitPacket() {
+	outPkt := &Packet{
+		IPAddr:  p.ipaddr,
+		Addr:    p.addr,
+		Seq:     p.sequence,
+		ReqTime: time.Now(),
+		Timeout: false,
+	}
+
+	p.waitPackets = append(p.waitPackets, outPkt)
```
Comment (sparrc) [inline ping.go]: Won't this list will grow forever? Can you please implement something to keep it short?

File: ping.go [hunk 2]
```diff
@@ -497,6 +513,34 @@ type IcmpData struct {
...
+
+	for _, waitPacket := range p.waitPackets {
+
+		timeDiff := time.Now().Sub(waitPacket.ReqTime)
+
+		if timeDiff.Seconds() > packetTimeoutSec && waitPacket.Timeout == false && outPkt.Nbytes <= 0 {
+			waitPacket.Timeout = true
+
+			timeoutHandler := p.OnTimeout
```
Comment (sparrc) [inline ping.go]: this assignment isn't necessary

File: ping.go [hunk 3]
```diff
@@ -497,6 +513,34 @@ type IcmpData struct {
 	Tracker int64
 }
 
+func (p *Pinger) checkWaitPacket() {
```
Comment (sparrc) [inline ping.go]: this function should return immediately if `p.OnTimeout == nil`

**Comment by CHTJonas** [comment]
Hi @eunchong - been a long time but are you interested in bringing this up-to-date and addressing the feedback? If not we'll close the PR.

**Comment by CHTJonas** [comment]
Poke @eunchong 😉

**Comment by CHTJonas** [comment]
Closing due to inactivity.

**Comment by yxxhero** [comment]
@CHTJonas Can it asign to me?

**Comment by CHTJonas** [comment]
IMO the solution to this PR, #165, #50 and #163 are the same and involve adding extra data/state to the `map[uuid.UUID]map[int]struct{}{}` type that we're using to keep track of packets locally. That anonymous `struct{}{}` would probably need to become something more well-defined.

IMO we don't want (too much) state in ICMP messages/headers if we can possibly help it. Trouble is that this doesn't yet exist in `master` yet so I think all of this is kinda blocked on #152. I think it would be best for you to jump on the #go-ping channel of the [Gophers Slack](https://invite.slack.golangbridge.org/) and discuss with us.

**Pull request closed without merging**